### PR TITLE
Bug Fix: Test Metrics Overwrite Validation Metrics

### LINKED ISSUE #669: Test set metrics overwrite validation set metrics in TensorBoard and are rejected for logging by Weights and Biases (W&B)
**Describe the bug**

At model train completion, the test set loss is written as iteration 0 to the TensorBoard / W&B chart `validation/lm_loss`, and the test set perplexity is written as iteration 0 to the chart `validation/lm_loss_ppl`. As the validation loss and perplexity has already been written to this chart, this results in TensorBoard deleting all the validation metrics, overwriting them with the test loss and perplexity values. W&B refuses to add the test metrics to the charts at all, throwing a warning that looks like `wandb: WARNING Step must only increase in log calls.  Step 0 < 32000; dropping {'validation/lm_loss': 1.715476632118225}.`

**To Reproduce**
1. Pip install and setup TensorBoard and W&B
2. Begin training a model with a train, validation, and test set
3. Observe in both TensorBoard and W&B that validation metrics are being logged
4. Allow the model to train to completion
5. Observe that the TensorBoard validation metrics are now gone, overwritten by the test set metrics
6. Observe the W&B error in the text logs / program output

**Expected behavior**
Test metrics should be written to their own charts.

**Proposed solution**
Test loss and perplexity should be written to their own charts `test/lm_loss` and `test/lm_loss_ppl` respectively.

**Screenshots**
image

**Environment (please complete the following information):**
 - GPUs: 4x A100 80 GB
- Configs: (configs that I used to reproduce the bug and test bug fixes are included below)

```
# GPT-2 pretraining setup
{
   # parallelism settings ( you will want to change these based on your cluster setup, ideally scheduling pipeline stages
   # across the node boundaries )
   "pipe-parallel-size": 1,
   "model-parallel-size": 1,

   # model settings
   "num-layers": 24,
   "hidden-size": 1024,
   "num-attention-heads": 16,
   "seq-length": 4096,
   "max-position-embeddings": 4096,
   "norm": "layernorm",
   "pos-emb": "rotary",
   "no-weight-tying": true,

   # these should provide some speedup but takes a while to build, set to true if desired
   "scaled-upper-triang-masked-softmax-fusion": false,
   "bias-gelu-fusion": false,



   # optimizer settings
   "optimizer": {
     "type": "Adam",
     "params": {
       "lr": 0.00003,
       "betas": [0.9, 0.999],
       "eps": 1.0e-8,
     }
   },
   "zero_optimization": {
    "stage": 1,
    "allgather_partitions": True,
    "allgather_bucket_size": 500000000,
    "overlap_comm": True,
    "reduce_scatter": True,
    "reduce_bucket_size": 500000000,
    "contiguous_gradients": True,
    "cpu_offload": False
  },
   # batch / data settings
   "train_micro_batch_size_per_gpu": 16,
   "data-impl": "mmap",
   "split": "949,50,1",

   # activation checkpointing
   "checkpoint-activations": true,
   "checkpoint-num-layers": 1,
   "partition-activations": true,
   "synchronize-each-layer": true,

   # regularization
   "gradient_clipping": 1.0,
   "weight-decay": 0.01,
   "hidden-dropout": 0,
   "attention-dropout": 0,

   # precision settings
   "fp16": {
     "fp16": true,
     "enabled": true,
     "loss_scale": 0,
     "loss_scale_window": 1000,
     "hysteresis": 2,
     "min_loss_scale": 1
   },

   # misc. training settings
   "train-iters": 100,
   "lr-decay-iters": 100,
   "distributed-backend": "nccl",
   "lr-decay-style": "constant",
   "warmup": 0.1,
   "save-interval": 25,
   "eval-interval": 25,
   "eval-iters": 10,

   # Checkpoint
   "finetune": true,

   # logging
   "log-interval": 10,
   "steps_per_print": 10,
   "keep-last-n-checkpoints": 4,
   "wall_clock_breakdown": true,
}
```

```
# Suggested data paths when using GPT-NeoX locally
{
  "train-data-paths": ["/mnt/4TBNVME/gpt-neox/data/preprocessed/train_text_document"],
  "test-data-paths": ["/mnt/4TBNVME/gpt-neox/data/preprocessed/test_text_document"],
  "valid-data-paths": ["/mnt/4TBNVME/gpt-neox/data/preprocessed/val_text_document"],

  "vocab-file": "/mnt/4TBNVME/gpt-neox/data/gpt2-vocab.json",
  "merge-file": "/mnt/4TBNVME/gpt-neox/data/gpt2-merges.txt",

  "save": "/mnt/4TBNVME/checkpoints_test",
  "load": "/mnt/4TBNVME/checkpoints_test",

  "checkpoint_validation_with_forward_pass": False,
  
  "tensorboard-dir": "/mnt/4TBNVME/logs/tensorboard/bug_fix_test",
  "log-dir": "/mnt/4TBNVME/logs/gptneox/bug_fix_test",

  "use_wandb": True,
  "wandb_host": "https://api.wandb.ai",
  "wandb_project": "neox_test"
}
```

```
# Add this to your config for sparse attention every other layer
{
  "attention_config": [[["local", "global"], "all"]],

  # sparsity config:
  # (these are the defaults for local sliding window sparsity, training will work without this here, but it's left in for
  # illustrative purposes)
  # see https://www.deepspeed.ai/tutorials/sparse-attention/#how-to-config-sparsity-structures for
  # more detailed config instructions and available parameters

  "sparsity_config": {
    "block": 16, # block size
    "num_local_blocks": 32,
  }
}
```

**Additional context**

I have a bug fix ready, will follow up with it.

### PR DESCRIPTION
**Pull request by pwstegman**
This PR fixes #669 where test set metrics overwrite validation set metrics on TensorBoard and aren't logged at all by Weights & Biases. It corrects the bug by writing test set metrics to their own charts, prefixed by `test/`. It also writes the metrics with the x-axis value being the last iteration, rather than iteration 0. This corrects the Weights & Biases error that the iteration (aka step) must always be increasing in subsequent log calls. For background, validation loss and perplexity are written to the charts `validation/lm_loss` and `validation/lm_loss_ppl`. At the end of training, the test loss and perplexity were also written to those two charts as iteration 0. This resulted in TensorBoard overwriting the validation data and in Weights & Biases throwing a warning such as `wandb: WARNING Step must only increase in log calls.  Step 0 < 32000; dropping {'validation/lm_loss': 1.715476632118225}.`

Tested manually to ensure that new charts were created for test metrics.

One thing I noticed was a comment on [line 147 of `megatron/training.py`](https://github.com/pwstegman/gpt-neox/commit/1c8c2ab3b858c9af25cf54d7f293e4907a31334b#diff-e2b248f8c422a601bcb0b7d93f96c1dff070f2694737e2b69f1def64ab9c1844L147) that said `iteration 0 in order to always use full test data`. However, this iteration of 0 is causing Weights & Biases to error out, as step must always be increasing in subsequent log calls. As can be seen in the called method beginning on [line 731](https://github.com/pwstegman/gpt-neox/blob/1c8c2ab3b858c9af25cf54d7f293e4907a31334b/megatron/training.py#L731), `iteration` is only used as part of the `tb_wandb_log` calls, and it does not appear to affect evaluation in any way. As such, I believe this comment may be outdated, and it should be safe to set the iteration to the final iteration value. I manually confirmed that setting the iteration value to the final iteration number fixed the Weights & Biases error.

### CODE REVIEW

**Comment by StellaAthena** [comment]
@pwstegman I know you were tinking with the tests... should we expect these tests to fail?

**Comment by pwstegman** [comment]
@StellaAthena Hmm, I wouldn't expect this PR to affect any of the checks, and I didn't make any changes to them in this PR.

Looks like the pre-commit check might be having an issue with black? For example, [this error](https://github.com/EleutherAI/gpt-neox/actions/runs/3040578756/jobs/4910059887#step:4:101) is an import error in black's `__init__.py`.

The update-documentation check seems to be throwing [an error](https://github.com/EleutherAI/gpt-neox/actions/runs/3040578756/jobs/4910059757#step:2:50) with git.

I'm not super experienced with GitHub checks, so not totally sure what's going on. Might not be related to the changes I made, since I didn't change anything that should have affected black or the documentation check.

**Review by StellaAthena — APPROVED** [review]
